crates/shared/src/event_handling.rs

```diff
@@ -69,51 +77,143 @@ where
         block_retriever: B,
         contract: C,
         store: S,
-        start_sync_at_block: Option<u64>,
+        start_sync_at_block: Option<BlockNumberHash>,
     ) -> Self {
         Self {
             block_retriever,
             contract,
             store,
-            last_handled_block: start_sync_at_block,
+            last_handled_blocks: {
+                match start_sync_at_block {
+                    Some(block) => vec![block],
+                    None => vec![],
+                }
+            },
         }
     }
 
     pub fn store(&self) -> &S {
         &self.store
     }
 
-    pub fn last_handled_block(&self) -> Option<u64> {
-        self.last_handled_block
+    pub fn last_handled_block(&self) -> Option<BlockNumberHash> {
+        self.last_handled_blocks.last().cloned()
     }
 
-    async fn event_block_range(&self) -> Result<RangeInclusive<BlockNumber>> {
-        // Instead of using only the most recent event block from the db we also store the last
-        // handled block in self so that during long times of no events we do not query needlessly
-        // large block ranges.
-        let last_handled_block = match self.last_handled_block {
-            Some(block) => block,
-            None => self.store.last_event_block().await?,
+    /// Defines block range, for which events should be fetched
+    /// Returns 3 parameters:
+    /// 1. Optional block number range for fetching reorg safe history (if it's needed)
+    /// 2. List of block numbers with hashes for fetching reorg unsafe blocks
+    /// 3. Bool indicator if reorg happened for reorg unsafe blocks
```

Comment:
Those pieces of documentation should now probably rather be moved to the fields of the struct.

Reply:
Ah yes, moved comment to struct.